framework/web/DbSession.php

```diff
@@ -169,14 +187,28 @@ public function writeSession($id, $data)
         // exception must be caught in session write handler
         // https://secure.php.net/manual/en/function.session-set-save-handler.php#refsect1-function.session-set-save-handler-notes
         try {
-            $fields = $this->composeFields($id, $data);
-            $fields = $this->typecastFields($fields);
-            $this->db->createCommand()->upsert($this->sessionTable, $fields)->execute();
+            // ensure backwards compatability (fixed #9438)
+            if ($this->writeCallback && !$this->fields) {
+                $this->fields = $this->composeFields();
+            }
+            // ensure data consistency
```

Comment:
Would you please elaborate what happens in this block?

Reply:
This is to ensure backwards compatibility & that unit tests won't fail.
Basically it forces populating custom fields on directly calling custom session handler `$session->writeSession()`.

Normally, custom fields are populated [just before closing the session](https://github.com/yiisoft/yii2/pull/17188/files#diff-6f6a191d70c6d117086ec1873e231af5R152).

But original unit tests would not call dedicated method `$session->close()` but instead directly custom handler method `$session->writeSession(sessId, sessDataString)` (because the session has not been opened), which IMO is not quite correct. 

I did not want to touch original unit tests, therefore this line must have been added as a work around solution. If direct injection of session data will drop in Yii3 these lines may be removed, or unit test should be rewritten to open/close session rather than directly calling custom handlers.